apps/chat/src/store/toolset/toolset.epics.ts

```diff
@@ -219,11 +241,188 @@ const updateToolsetEpic: AppEpic = (action$, _state, { router }) =>
     }),
   );
 
+const getBookmarkedToolsetsEpic: AppEpic = (action$, state$) =>
+  action$.pipe(
+    ofType(ToolsetActions.getBookmarkedToolsets.type),
+    switchMap(() => {
+      const allToolsets = ToolsetSelectors.selectToolsets(state$.value);
+
+      const myToolsetsReferences = allToolsets
+        .filter((toolset) => isMyEntity(toolset) || toolset.sharedWithMe)
+        .map((myToolset) => myToolset.reference);
+
+      return ClientDataService.getBookmarkedToolsets().pipe(
+        switchMap((bookmarkedToolsets) => {
+          if (!bookmarkedToolsets) {
+            return of(
+              ToolsetActions.getBookmarkedToolsetsFail(myToolsetsReferences),
+            );
+          }
+
+          const actions: Observable<AppAction>[] = [];
+
+          const bookmarkedToolsetsSet = new Set(bookmarkedToolsets);
+
+          const references = [...bookmarkedToolsets, ...myToolsetsReferences];
```

Comment:
```suggestion
          const references = uniq([...bookmarkedToolsets, ...myToolsetsReferences]);
```

to avoid double call of the `addBookmarkedToolsets` action?
Or it's fine here?

Reply:
It's ok here since all already installed toolsets filtered from references at line 268.